### PR TITLE
GSOC 2020: Port bridge / irc modules to async / await, remove futures 0.1 dependencies

### PR DESCRIPTION
**Pull request by VanillaBrooks**

### CODE REVIEW

**Comment by jplatte** [comment]
@VanillaBrooks Sorry if this wasn't clear from our communication on Matrix, but I'm still waiting for you to rebase this.

File: src/matrix/mod.rs [hunk 1]
```diff
@@ -84,41 +83,59 @@ impl MatrixClient {
...
             .expect("expected host in base_url")
             .to_string();
-        let port = base_url.port_or_known_default().unwrap();
-        let tls = match base_url.scheme() {
+        let _port = base_url.port_or_known_default().unwrap();
+        let _tls = match base_url.scheme() {
             "http" => false,
             "https" => true,
             _ => panic!("Unrecognized scheme {}", base_url.scheme()),
         };
```
Comment (VanillaBrooks) [inline src/matrix/mod.rs]: previously these were used with the custom http implementation, but since `url::Url` and is later converted to `hyper::Uri`, i think these are safe to remove.

File: src/matrix/mod.rs [hunk 2]
```diff
@@ -156,10 +177,7 @@ impl MatrixClient {
         Ok(json_response)
     }
 
-    pub async fn join_room(
-        &mut self,
-        room_id: &str,
-    ) -> Result<protocol::RoomJoinResponse, Error> {
+    pub async fn join_room(&mut self, room_id: &str) -> Result<protocol::RoomJoinResponse, Error> {
```
Comment (VanillaBrooks) [inline src/matrix/mod.rs:180]: A lot of these function signatures are changing from rustfmt and I dont know why

File: src/matrix/mod.rs [hunk 3]
```diff
@@ -204,12 +222,15 @@ impl MatrixClient {
         }
     }
 
-    fn poll_sync(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Result<protocol::SyncResponse, Error>>> {
+    fn poll_sync(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context,
+    ) -> Poll<Option<Result<protocol::SyncResponse, Error>>> {
```
Comment (VanillaBrooks) [inline src/matrix/mod.rs:228]: same here

File: src/irc/transport.rs [hunk 4]
```diff
@@ -57,74 +67,80 @@ impl<S: AsyncRead + AsyncWrite> IrcServerConnection<S> {
...
             }
-
-            if let Some(ref t) = inner.write_notify {
-                t.notify();
-            }
         }
-        self.poll_write().ok();
+
+        // ignore the result of the poll
+        self.write().await.ok();
```
Comment (jplatte) [inline src/irc/transport.rs]: Instead of adding that comment, could you rewrite this and any other occurrences of the `<expr>.ok();` pattern to `let _ = <expr>;`, which is the standard way of saying "ignore result of <expr>" in Rust?

File: src/matrix/sync.rs [hunk 5]
```diff
@@ -47,10 +47,10 @@ enum RequestStatus {
...
-        let host = base_url.host_str().expect("expected host in base_url");
-        let port = base_url.port_or_known_default().unwrap();
+        let _host = base_url.host_str().expect("expected host in base_url");
+        let _port = base_url.port_or_known_default().unwrap();
 
-        let tls = match base_url.scheme() {
+        let _tls = match base_url.scheme() {
             "http" => false,
             "https" => true,
             _ => panic!("Unrecognized scheme {}", base_url.scheme()),
```
Comment (VanillaBrooks) [inline src/matrix/sync.rs]: These can also probably be removed

File: src/stream_fold.rs [hunk 6]
```diff
@@ -12,65 +12,84 @@
...
 #[must_use = "futures do nothing unless polled"]
-pub struct StreamFold<I, E, S: Stream<Item = I, Error = E>, V, F: FnMut(I, V) -> (bool, V)> {
-    func: F,
-    state: StreamFoldState<S, V>,
+pub struct StreamFold<S, W, T, V>
+where
+    S: Stream<Item = Result<T, V>>,
+    W: StateUpdate<Result<T, V>>,
+{
+    //stream: Mutex<Pin<Box<S>>>,
```
Comment (VanillaBrooks) [inline src/stream_fold.rs]: remove this

File: src/main.rs [hunk 7]
```diff
@@ -22,23 +22,22 @@ extern crate slog;
...
-use futures::Future;
+use futures::stream::StreamExt;
 
 use slog::Drain;
 
 use std::cell::RefCell;
 use std::fs::File;
 use std::io::{self, Read};
 use std::net::SocketAddr;
+
```
Comment (jplatte) [inline src/main.rs]: Extra newline seems pointless.

File: src/main.rs [hunk 8]
```diff
@@ -22,23 +22,22 @@ extern crate slog;
...
 
-use tokio_core::net::TcpListener;
-use tokio_core::reactor::Core;
+use tokio::net::TcpListener;
 
-use native_tls::{Identity, TlsAcceptor};
+use native_tls::Identity;
 
-use tasked_futures::TaskExecutor;
+use std::task::Context;
```
Comment (jplatte) [inline src/main.rs]: Please move this up to the other `std` imports.

File: src/main.rs [hunk 9]
```diff
@@ -49,9 +48,9 @@ lazy_static::lazy_static! {
     };
 }
 
-futures::task_local! {
+std::thread_local! {
```
Comment (jplatte) [inline src/main.rs]: Why change the task local into a thread local?

File: src/main.rs [hunk 10]
```diff
@@ -142,42 +142,44 @@ fn main() {
...
         };
 
-        let new_handle = handle.clone();
+        //let new_handle = handle.clone();
 
         let cloned_ctx = ctx.clone();
 
         // Set up a new task for the connection. We do this early so that the logging is correct.
-        let setup_future = futures::lazy(move || {
+        let setup_future = futures::future::lazy(move |_cx: &mut Context| {
```
Comment (jplatte) [inline src/main.rs]: Would be good for readability to shorten `futures::future` into just `future` by `use`ing `futures::future` at the start of the file.

File: src/matrix/mod.rs (hunk 1 above)
Comment (jplatte) [reply src/matrix/mod.rs]: Yeah, sounds good.

File: src/stream_fold.rs [hunk 11]
```diff
@@ -12,65 +12,84 @@
...
-use futures::{Async, Future, Poll};
+use futures::task::{Context, Poll};
 
-use std::mem;
+use std::boxed::Box;
 
-/// A Stream adapater, similar to fold, that consumes the start of the stream to build up an
-/// object, but then returns both the object *and* the stream.
+use std::pin::Pin;
+
```
Comment (jplatte) [inline src/stream_fold.rs]: Why extra newlines between `std` imports?

File: src/matrix/mod.rs (hunk 2 above)
Comment (jplatte) [reply src/matrix/mod.rs:180]: Potentially the codebase just wasn't `rustfmt`ed before. I've thought this for some other bits of code. Don't worry about it.

File: src/main.rs [hunk 12]
```diff
@@ -142,42 +142,44 @@ fn main() {
...
+
         let peer_log = log.new(o!("ip" => format!("{}", addr.ip()), "port" => addr.port()));
 
         let ctx = ConnectionContext {
             logger: Arc::new(peer_log),
             peer_addr: addr,
         };
 
-        let new_handle = handle.clone();
+        //let new_handle = handle.clone();
```
Comment (VanillaBrooks) [inline src/main.rs]: remove

File: src/irc/mod.rs [hunk 13]
```diff
@@ -46,94 +48,105 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
...
+
+        match new_item {
+            IrcCommand::Nick { nick } => self.nick = Some(nick),
+            IrcCommand::User { user, real_name } => {
+                self.user = Some(user);
+                self.real_name = Some(real_name);
+            }
+            IrcCommand::Pass { password } => self.password = Some(password),
+            _c => {
+                //debug!(ctx.logger, "Ignore command during login"; "cmd" => c.command());
```
Comment (VanillaBrooks) [inline src/irc/mod.rs]: add this back

File: src/bridge/mod.rs [hunk 14]
```diff
@@ -253,42 +248,65 @@ impl<IS: AsyncRead + AsyncWrite + 'static> Bridge<IS> {
...
 
-    fn poll_matrix(&mut self) -> Poll<(), io::Error> {
-        loop {
-            if let Some(sync_response) = futures::try_ready!(self.matrix_client.poll()) {
-                self.handle_sync_response(sync_response);
-            } else {
-                self.closed = true;
-                self.stop();
-                return Ok(Async::Ready(()));
-            }
```
Comment (VanillaBrooks) [inline src/bridge/mod.rs:282]: I _believe_ `.await` will replicate how the `self.stop()` call to `taskedfutures` work based on eric's explanation, but this needs to be tested

File: src/bridge/mod.rs [hunk 15]
```diff
@@ -253,42 +248,65 @@ impl<IS: AsyncRead + AsyncWrite + 'static> Bridge<IS> {
...
+    pub async fn poll_matrix(&mut self) -> Result<(), Error> {
+        while let Some(response) = self.matrix_client.as_mut().next().await {
+            let response = response?;
+            self.handle_sync_response(response).await;
```
Comment (VanillaBrooks) [inline src/bridge/mod.rs:285]: I believe `.await` will replicate how the `self.stop()` call to `taskedfutures` work based on eric's explanation but this needs to be tested before merging to master

File: src/bridge/mod.rs [hunk 16]
```diff
@@ -253,42 +248,65 @@ impl<IS: AsyncRead + AsyncWrite + 'static> Bridge<IS> {
...
 }
 
-impl<IS: AsyncRead + AsyncWrite> TaskExecutor for Bridge<IS> {
-    type Error = io::Error;
-
-    fn task_executor_mut(&mut self) -> &mut TaskExecutorQueue<Self, io::Error> {
-        &mut self.executor_queue
+quick_error! {
+    #[derive(Debug)]
+    pub enum Error {
```
Comment (VanillaBrooks) [inline src/bridge/mod.rs:293]: Some of these are probably unnecessary, ill have to check since this was a copy paste from the matrix module

File: src/main.rs (hunk 9 above)
Comment (VanillaBrooks) [reply src/main.rs]: I mentioned this on matrix but no one really got back to me on it. Ill copy paste it here so its more open:

since `futures::task_local!` does not exist in `futures` 0.3 I'm using `tokio::task_local!`, but there is one notable difference in that you no longer set the default value:

```
futures::task_local!{
    static CONTEXT: RefCell<Option<ConnectionContext>> = RefCell::new(None);
}
```
becomes:

```
tokio::task_local!{
    static CONTEXT: RefCell<Option<ConnectionContext>>;
}
```
Running the tests from my branch here gives the error:

```
'cannot access a Task Local Storage value without setting it via `LocalKey::set`: AccessError'
```
but I dont see a `LocalKey::set` anywhere in the `tokio` docs. It looks like in [the source code](https://docs.rs/tokio/0.2.21/src/tokio/task/task_local.rs.html#34-46) `tokio` uses `std::thread_local!`, is it safe to use this as a replacement?

File: src/matrix/mod.rs (hunk 2 above)
Comment (VanillaBrooks) [reply src/matrix/mod.rs:180]: Ill look into adding it to the ci

File: src/irc/mod.rs [hunk 17]
```diff
@@ -46,94 +48,105 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
...
-                let nick = user_nick.nick.expect("nick");
-
-                let password = if let Some(p) = user_nick.password {
-                    p
-                } else {
-                    irc_conn.write_password_required(&nick);
-                    return Err(io::Error::new(
-                        io::ErrorKind::InvalidData,
-                        "IRC user did not supply password",
-                    ));
```
Comment (VanillaBrooks) [inline src/irc/mod.rs:103]: This did not get ported since I check that all fields are filled before exiting the `StreamFold`, whereas this version allows the password field to stay `None` until later. Will update for this

File: Cargo.toml [hunk 18]
```diff
@@ -22,23 +19,15 @@ serde_json = "1.0.53"
...
 
 [dependencies.openssl]
 version = "0.10.0"
 
-[dependencies.tasked-futures]
-git = "https://github.com/erikjohnston/tasked-futures.git"
-optional = false
-
 [dependencies.tokio-timer]
 git = "https://github.com/tokio-rs/tokio-timer"
```
Comment (phlmn) [inline Cargo.toml]: Is this still used? We certainly want to get rid of it.

File: src/bridge/mod.rs [hunk 19]
```diff
@@ -16,146 +16,129 @@
...
-                            bridge.joining_map.insert(room_id, channel);
-                        }
-                    });
-                // TODO: Handle failure of join. Ensure that joining map is cleared.
-                self.spawn(join_future);
+                let join_future =
+                    if let Ok(response) = self.matrix_client.join_room(channel.as_str()).await {
+                        response
+                    } else {
+                        // TODO: log this
```
Comment (phlmn) [inline src/bridge/mod.rs:126]: yeah, logging here would be really nice

File: src/main.rs [hunk 20]
```diff
@@ -142,42 +142,44 @@ fn main() {
...
 
     info!(log, "Started listening"; "addr" => bind_addr, "tls" => tls);
 
     // This is the main loop where we accept incoming *TCP* connections.
-    let done = socket.incoming().for_each(move |(socket, addr)| {
+    while let Some(Ok(tcp_stream)) = socket.next().await {
+        let addr = if let Ok(addr) = tcp_stream.peer_addr() {
+            addr
+        } else {
+            // TODO: log this
```
Comment (phlmn) [inline src/main.rs]: Could you add a simple log here?

File: src/main.rs [hunk 21]
```diff
@@ -189,47 +191,68 @@ fn main() {
...
+                        .await
+                        .unwrap();
+                loop {
+                    if let Err(_) = bridge.poll_irc().await {
+                        break;
+                        // TODO: log this
+                    }
+                    if let Err(_) = bridge.poll_matrix().await {
+                        break;
+                        // TODO: log this
```
Comment (phlmn) [inline src/main.rs]: Just add simple logs here, then it's done :)

File: src/matrix/mod.rs [hunk 22]
```diff
@@ -334,6 +350,8 @@ mod tests {
 
     #[tokio::test]
     async fn send_text_message() {
+        //crate::CONTEXT.scope(std::cell::RefCell::new(None),  async {}).await;
```
Comment (phlmn) [inline src/matrix/mod.rs]: What is this doing here?

**Review by phlmn — CHANGES_REQUESTED** [review]
I really like where this is going 🙂 

Have all files been formatted now? I find it quite irritating to have all those formatting changes in the reviews. Maybe we should make a dedicated PR just for formatting, where we can just skim the changes, since we know it's just formatting.

File: src/matrix/mod.rs (hunk 22 above)
Comment (VanillaBrooks) [reply src/matrix/mod.rs]: This is a method to set the state for `crate::CONTEXT` when using `tokio::task_local!` as a replacement for `futures::task_local!` since the `tokio` version doesn't provide default state. [docs here](https://docs.rs/tokio/0.2.21/tokio/task/struct.LocalKey.html#method.scope).  I never actually got this to work, so i switched to `std::thread_local!`. There's an unresolved discussion about this above, but for now ill remove this dead code.

File: Cargo.toml (hunk 18 above)
Comment (VanillaBrooks) [reply Cargo.toml]: Not used, good catch

File: src/bridge/mod.rs [hunk 23]
```diff
@@ -98,11 +98,12 @@ impl<IS: AsyncRead + AsyncWrite + 'static + Send> Bridge<IS> {
             IrcCommand::PrivMsg { channel, text } => {
                 if let Some(room_id) = self.mappings.channel_to_room_id(&channel) {
                     info!(self.ctx.logger, "Got msg"; "channel" => channel.as_str(), "room_id" => room_id.as_str());
-                    self.matrix_client
+
+                    let _ = self
```
Comment (jplatte) [inline src/bridge/mod.rs]: Oh, sorry, I guess I misread this. If we actually do something on `Err` (run `task_warn` in this case), that shouldn't be a `map_err` at all, instead you can use `if self.matrix_client.send_text_message(room_id, text).await.is_err() { task_warn!("Failed to send"); }`.

File: src/irc/mod.rs [hunk 24]
```diff
@@ -47,48 +47,61 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
...
-#[derive(Debug, Clone, Default)]
+#[derive(Clone)]
 struct UserNickBuilder {
+    ctx: ConnectionContext,
     nick: Option<String>,
     user: Option<String>,
     real_name: Option<String>,
     password: Option<String>,
 }
+impl UserNickBuilder {
```
Comment (jplatte) [inline src/irc/mod.rs]: Please always leave an empty line between toplevel items.

File: src/irc/mod.rs [hunk 25]
```diff
@@ -47,48 +47,61 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
...
     fn to_user_nick(self) -> UserNick {
         UserNick {
             nick: self.nick.unwrap(),
             user: self.user.unwrap(),
             real_name: self.real_name.unwrap(),
-            password: self.password.unwrap(),
+            password: self.password,
         }
     }
+    /// Check that the user has input all fields except password.
```
Comment (jplatte) [inline src/irc/mod.rs]: Same here.

```suggestion
    }

    /// Check that the user has input all fields except password.
```

File: src/irc/mod.rs [hunk 26]
```diff
@@ -47,48 +47,61 @@ pub struct IrcUserConnection<S: AsyncRead + AsyncWrite> {
...
+    ///
+    /// A missing password will be caught in `await_login`.
```
Comment (jplatte) [inline src/irc/mod.rs:82]: Is this change in behavior somehow linked to porting to async/await? If not, please do this in a separate PR, this one is already very large.

File: src/irc/mod.rs [hunk 27]
```diff
@@ -119,22 +132,39 @@ where
 
         let ctx_clone = ctx.clone();
 
-        let folder = crate::stream_fold::StreamFold::new(irc_conn, UserNickBuilder::default());
+        let folder =
+            crate::stream_fold::StreamFold::new(irc_conn, UserNickBuilder::with_context(ctx));
+
+        // We cant consusme the future since we need to split it into the irc connection and
```
Comment (jplatte) [inline src/irc/mod.rs]:
```suggestion
        // We cant consume the future since we need to split it into the irc connection and
```

File: src/irc/mod.rs (hunk 26 above)
Comment (VanillaBrooks) [reply src/irc/mod.rs:82]: Im not sure what you mean. The changes regarding getting IRC user information from the irc connection stream required a large rewrite of `stream_fold.rs` + the creation of `stream_fold::StateUpdate` trait to be compatible with the `Pin<&mut Self>` requirements of futures 0.3.

As far as checking the password field in `await_login` (and allowing the `StreamFold` to keep it as `None` in the builder): this is an addition that I made recently to mirror the way futures 0.1 does it after spotting the discrepancy in the diffs:

https://github.com/matrix-org/matrix-ircd/blob/7d5f660e11037ee380a08ed74cb6ea95dba3bb9d/src/irc/mod.rs#L96-L104

Comment (VanillaBrooks) [reply src/irc/mod.rs:82]: How would you like to proceed?

File: src/bridge/mod.rs (hunk 23 above)
Comment (VanillaBrooks) [reply src/bridge/mod.rs]: I agree, updated it

File: src/irc/mod.rs (hunk 27 above)
Comment (VanillaBrooks) [reply src/irc/mod.rs]: thanks, good catch

File: src/irc/mod.rs (hunk 26 above)
Comment (jplatte) [reply src/irc/mod.rs:82]: Sounds like it is indeed related, so consider this resolved.

File: src/macros.rs [hunk 28]
```diff
@@ -14,52 +14,46 @@
...
+// $ctx in all macro arguments is of type crate::ConnectionContext
 
-        CONTEXT.with(|m| {
-            if let Some(ref ctx) = m.borrow().as_ref() {
-                log!(ctx.logger.as_ref(), $lvl, "", $($args)+)
-            } else {
-                log!(crate::DEFAULT_LOGGER, $lvl, "", $($args)+)
-            }
-        });
+macro_rules! task_log {
```
Comment (jplatte) [inline src/macros.rs:19]: This doesn't seem to be very useful now that the context is no longer a thread local. As a followup (after this PR), could you rewrite all uses of it to the regular log macros?

File: src/main.rs [hunk 29]
```diff
@@ -168,16 +163,16 @@ async fn main() {
             peer_addr: addr,
         };
 
-        let cloned_ctx = ctx.clone();
+        //let cloned_ctx = ctx.clone();
```
Comment (jplatte) [inline src/main.rs]: Please don't leave this commented code lying around. If it's no longer needed, remove it.

File: src/macros.rs (hunk 28 above)
Comment (VanillaBrooks) [reply src/macros.rs:19]: I agree thats a good change, mildly confusing how log macros take `ctx.logger.as_ref()` but the task_log just take `ctx`

File: src/main.rs (hunk 20 above)
Comment (jplatte) [reply src/main.rs]: @VanillaBrooks Why did you mark this as resolved? There is still no logging here.

File: src/main.rs (hunk 12 above)
Comment (jplatte) [reply src/main.rs]: There's still commented-out code here.

File: src/irc/mod.rs (hunk 13 above)
Comment (jplatte) [reply src/irc/mod.rs]: Why did you mark as resolved?

**Review by jplatte — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 12 above)
Comment (VanillaBrooks) [reply src/main.rs]: I dont know what to tell you. Its not there on my branch https://github.com/VanillaBrooks/matrix-ircd/blob/async-await/src/main.rs#L170-L183. ctrl + F `new_handle` shows nothing as well.

File: src/irc/mod.rs (hunk 13 above)
Comment (VanillaBrooks) [reply src/irc/mod.rs]: This one is also fixed on my branch but is not carrying over correctly (?) https://github.com/VanillaBrooks/matrix-ircd/blob/async-await/src/irc/mod.rs#L103-L106. Any idea whats going on here ?

File: src/main.rs (hunk 21 above)
Comment (VanillaBrooks) [reply src/main.rs]: These logs appear on my branch https://github.com/VanillaBrooks/matrix-ircd/blob/async-await/src/main.rs#L231-L247. more logging is in the tls bridge version above.

File: src/main.rs (hunk 20 above)
Comment (VanillaBrooks) [reply src/main.rs]: These I did actually mess up on. Should have paid more attention to the line numbers! Ive added it now.

File: src/bridge/mod.rs (hunk 19 above)
Comment (jplatte) [reply src/bridge/mod.rs:126]: @VanillaBrooks This still needs to be addressed.

File: src/main.rs [hunk 30]
```diff
@@ -189,47 +187,68 @@ fn main() {
...
+                task_info!(ctx, "Finished");
+            });
 
             // We spawn the future off, otherwise we'd block the stream of incoming connections.
             // This is what causes the future to be in its own chain.
-            handle.spawn(future);
+            //
+            // These are spawn_local due to it not requiring spawn_fut to be Send, but it could
+            // possibly be `spawn` in the future after changing trait bounds.
+            tokio::task::spawn_local(spawn_fut);
```
Comment (jplatte) [inline src/main.rs:229]: [This function panics if called outside of a local task set.](https://docs.rs/tokio/0.2.22/tokio/task/fn.spawn_local.html) You can't use it like this.

Comment (jplatte) [reply src/main.rs:229]: I'm guessing this also got lost in my not-actually-submitted previous review.

**Review by jplatte — CHANGES_REQUESTED** [review]

**Comment by jplatte** [comment]
Merged even with some outstanding issues because there seems to be something wrong with the diff and adding more to this PR isn't going to make further review easier.

File: src/main.rs (hunk 30 above)
Comment (VanillaBrooks) [reply src/main.rs:229]: This means that if `spawn_local` is called outside of a task it will panic? So `tokio::spawn` would be required, and therefore everything must be `Send`?

Comment (jplatte) [reply src/main.rs:229]: No. Everything is inside a task with `#[tokio::main]`. What this is talking about it `LocalTaskSet`. Please look at the linked documentation (and the one for `LocalTaskSet`) if you want to know more. We probably should be making everything `Send` and not use `LocalTaskSet` though.
